lib/Clam/Properties/BndCheck.cc

```diff
@@ -0,0 +1,338 @@
+    } else {
+      CRAB_LOG("cfg-bnd-check",
+               llvm::errs()
+                   << "load: check size and offset are in bounds: " << I
+                   << "\n");
+      var_t ptr_obj_sz = it->second.first;
+      var_t ptr_offset = it->second.second;
+      Type *elemTy = key->getType()->getPointerElementType();
+      const llvm::DataLayout *dl = getInsDatalayout(I);
+      unsigned size_of = dl->getTypeSizeInBits(elemTy) / 8;
+      addBndCheckAssertion(s.getRegion(), ptr_obj_sz, ptr_offset, bb, I, size_of);
+      crab_lit_ref_t lhs = m_lfac.getLit(I);
+      if (lhs->getVar().get_type().is_reference()) {
+        var_t ref_obj_sz = m_lfac.mkIntVar(m_width);
+        // obtain an overappriximation of deref
+        // using havoc to initialize size and offset
+        bb.havoc(ref_obj_sz, "approx size when loading a ptr of ptr");
+        var_t ref_offset = m_lfac.mkIntVar(m_width);
+        bb.havoc(ref_offset, "approx offset when loading a ptr of ptr");
+
+        // add pair to the map by llvm pointer
+        Value *ref_v = llvm::cast<llvm::Value>(&I);
+        m_ref_bnd_map->insert({ref_v, {ref_obj_sz, ref_offset}});
+      }
+    }
+  }
+
+  // -- alloc site ptr bound utilities construction
+  // From LLVM IR to crab IR we perform the following transformation:
+  // p := alloc(sz) => p := make_ref(region, sz)
+  // We use the idea from fat pointer like by adding two additional variables:
+  // p_offset := 0;
+  // p_size := sz;
+  // where offset indicates the offset of current pointer
+  // the size means the total size of allcation
+  void visitAfterAlloc(llvm::Instruction &I, const llvm::TargetLibraryInfo &tli,
+                       CrabMakeRefOps &s) {
+    // llvm::errs() << I << "\n";
+    // llvm::errs() << *I.getOperand(0) << "\n";
+    basic_block_t &bb = s.getBasicBlock();
+    CRAB_LOG("cfg-bnd-check",
+             llvm::errs() << "alloc: add size and offset for " << I << "\n");
+    llvm::Value *op;
+    unsigned sz;
+    if (isa<AllocaInst>(I)) {
+      AllocaInst *allocaI = dyn_cast<AllocaInst>(&I);
+      op = allocaI->getArraySize();
+      const llvm::DataLayout *dl = getInsDatalayout(*allocaI);
+      sz = dl->getTypeSizeInBits(allocaI->getAllocatedType()) / 8;
+    } else {
+      op = I.getOperand(0);
+    }
+    llvm::Value &operand = *op;
+    crab_lit_ref_t opd = m_lfac.getLit(operand);
+    if (!opd->isInt()) {
+      CLAM_WARNING("Unexpected type of alloc size operand");
+      return;
+    }
+    assert(opd->isInt());
+    auto Iopd = std::static_pointer_cast<const crabIntLit>(opd);
+    unsigned width = Iopd->getBitwidth();
+    var_t tmp_obj_sz = m_lfac.mkIntVar(width);
+    if (isa<AllocaInst>(I))
+      bb.assign(tmp_obj_sz, m_lfac.getExp(opd) * sz);
+    else
+      bb.assign(tmp_obj_sz, m_lfac.getExp(opd));
+    var_t ref_obj_sz = m_lfac.mkIntVar(m_width);
+    if (m_width != width) // convert assigned size to fixed width 64
```

Comment:
Ideally, I wish to use width from what we get either from DataLayout or from the method of `crabIntLit`. However, the width of offset and size variables should always remain consistent. The reason I did this is that there are several conditions that I cannot make consistent width. For instance:
```c++
uint16_t *ptr;
uint16_t element = 10;
uint64_t len = uint64_t_nd();
__CRAB_assume(len >= 0);
__CRAB_assume(len < 10);
ptr = malloc(sizeof(uint16_t) * len);
// ptr_offset = 0
// ptr_size = sizeof(uint16_t) * len
// where width of sizeof(uint16_t) * len may vary
uint16_t *tmp;
if (len == 1) {
  tmp = &element;
}
else if (len == 2){
  tmp = ptr + 4;
}
else {
  tmp = NULL;
}
// suppose we translate above as phi_node in LLVM IR 
// which will be three gep ops in CRAB IR
// The width for element part and ptr part will be inconsistent, 
    where offset and size of ptr has width of size_t
    where offset and size of element could be choosen 
    from datalayout which will be width of uint16_t
// it will cause the problem when we do the offset's and size's assignments for the tmp
```

Reply:
I understand the problem of making sure that all the integer variables used for ghost offsets and sizes have consistent bitwidth. But not sure about the solution. 

First, I don't like is this line:
```
const unsigned m_width = 64; // Fixed width for vars of size and offset
```

I think we could use the function 

```
unsigned getMaxPointerSizeInBits() const;
```

defined in DataLayout.  

Second, the code that translates an allocation function or alloca instruction shouldn't have any inconsistency since you could use  always `getMaxPointerSizeInBits`. Similarly, the code that generates assertions (Load/Store, etc) shouldn't have any inconsistency because you can also use `getMaxPointerSizeInBits`. 
However, the code that might produce some type inconsistency is the translation to Crab `gep_ref`. This translation already generates a numerical offset. We want to use directly that offset for our instrumentation. We would need to make sure that the numerical offset is using the same bitwidth that `getMaxPointerSizeInBits`. Right now I think it doesn't so we would need to add (optionally) and signed extension. 

Btw, related to the translation of LLVM GEP instructions. Let me just explain what we do to give more context. In general, each GEP operand can be a pointer to a different kind of integer. This is problematic to the translation to Crab because it might generate many intermediate arithmetic operations (basically one per GEP operand). Each intermediate arithmetic operation might have different bitwidth. So the solution I took was to first do a scan over all GEP operands and collect all possible bitwidths. Then, I assume that all intermediate arithmetic operations use the maximum bitwidth. If they don't then I signed extend. In SeaHorn (e.g., ufo semantics) this is not a problem because we are not limited to three-address code so we just generate a single arithmetic operation that models the value of the offset.

Reply:
Perhaps I can also use `getMaxPointerSizeInBits` in the translation of GEP so then, we shouldn't have any inconsistency while instrumenting the code with ghost offset/size variables.

Reply:
@yusen: can you try to replace this function 

https://github.com/seahorn/clam/blob/master/lib/Clam/CfgBuilder.cc#L928

with 

```
unsigned
CrabIntraBlockBuilder::getMaxBitWidthFromGepIndexes(GetElementPtrInst &I) {
  return m_dl->getMaxPointerSizeInBits();
}
```

And then use elsewhere `getMaxPointerSizeInBits()` rather than the hardcoded 64 value?

I think these changes should get rid completely of the type consistency problems.

Reply:
Sure. Let me try it. If that works. I will use it as my solution. However, the `zext` or `trunc` might be used for ghost variables at the allocation site.